### ISSUE TITLE
Non-Deterministic results for a custom user-agent string

### ISSUE THREAD

**Issue by coladarci** [elixytics/ua_inspector#17]
Hey all - we are using this library and one of the UserAgents we are inspecting is a string being generated by a native ios app. Our engineers put together a string that is parasable by the client by following the rules of creating a custom user agent string. 

I'm not expecting for you to help support custom strings as that's obviously out of scope for this library. What I **am** interested in your help w/ is that this custom string is parsed differently each time it is run; like something is non-deterministic about the algorithm. See this test, w/ the same seed; the first time it works, the second time it fails:

```
Randomized with seed 244484
mix test test/peek/actor_test.exs --seed 244484
...

Finished in 0.1 seconds
2 doctests, 1 tests, 0 failures

Randomized with seed 244484
mix test test/peek/actor_test.exs --seed 244484
.

  1) test ua test (Peek.ActorTeset)
     test/peek/actor_test.exs:52
     match (=) failed
     code:  assert %{device: %{type: "tablet"}} = UAInspector.parse_client("Peek Pro/5.11.0 (iPad; iOS 12.2; Scale/2.00) (iPad7,1;)")
     right: %UAInspector.Result{
              client: %UAInspector.Result.Client{
                engine: :unknown,
                engine_version: :unknown,
                name: "Peek Pro",
                type: "mobile app",
                version: "5.11.0"
              },
              device: %UAInspector.Result.Device{
                brand: "Apple",
                model: :unknown,
                type: :unknown
              },
              os: %UAInspector.Result.OS{
                name: "iOS",
                platform: :unknown,
                version: "12.2"
              },
              user_agent: "Peek Pro/5.11.0 (iPad; iOS 12.2; Scale/2.00) (iPad7,1;)"
            }
     stacktrace:
       test/peek/actor_test.exs:53: (test)

..

Finished in 0.1 seconds
2 doctests, 1 tests, 1 failure
```

The test, itself, is simple:
```
assert %{device: %{type: "tablet"}} = UAInspector.parse_client("Peek Pro/5.11.0 (iPad; iOS 12.2; Scale/2.00) (iPad7,1;)")
```

Any ideas here?

Thanks!

**Comment by mneudert** [comment]
As a quick guess I would say the database was not fully loaded in one of your test runs.

You can try to add some waiting cycles to await the loading by using `UAInspector.ready?` like I had to add for the parser verification: https://github.com/elixir-inspector/ua_inspector/blob/master/verify/lib/mix/tasks/ua_inspector/verify.ex#L152

**Comment by coladarci** [comment]
This fixed it.. thanks for the tip!

That said, a little smelly, no? Is there not a way this library can block higher up the chain to ensure the DB is loaded before continuing?

**Comment by mneudert** [comment]
Yeah, it would probably be a really really good idea to at least provide a "known working" `wait_until_ready` function for a more deterministic startup (and state).

With that in place `UAInspector.Application.start/2` could block until the database is fully loaded or any depending application can wait until the loading process is done. As `ready?` only checks a bunch of ETS tables it should also be possible to have the parsing function itself wait until everything is good to go.

Might need some tweaking if a database fails to parse (and therefore not receiveing a ready state) but that should be possible with a just slightly more intelligent detection.

**Comment by coladarci** [comment]
Yeah we use a handful of libraries who have the same challenge and all seemed to have solved it. An example is https://github.com/elixir-geolix/geolix - might be worth seeing how they handle this.

**Comment by mneudert** [comment]
The just released `v1.1.0` has a new configuration option to ensure a synchronous startup is performed:

```elixir
config :ua_inspector,
  startup_sync: true
```

In the next (major) version this will most likely become the default behaviour.

Thank you for making me fix that problem :D :heart: :blue_heart: :green_heart: :purple_heart: :yellow_heart:
